File: commands/battle.js
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const { gameManager } = require('../game/gameManager');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('battle')
        .setDescription('Initiate or manage a battle.')
        .addSubcommand(subcommand =>
            subcommand
                .setName('start')
                .setDescription('Start a battle on a tile where opposing armies are present.')
                .addIntegerOption(option => option.setName('x').setDescription('The x-coordinate of the battle tile.').setRequired(true))
                .addIntegerOption(option => option.setName('y').setDescription('The y-coordinate of the battle tile.').setRequired(true)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('place')
                .setDescription('Place a unit on the battlefield during the placement phase.')
                .addStringOption(option =>
                    option.setName('unit_type')
                        .setDescription('The type of unit to place.')
                        .setRequired(true)
                        .addChoices(
                            { name: 'Infantry', value: 'infantry' },
                            { name: 'Commander', value: 'commander' },
                            { name: 'Shock', value: 'shock' },
                            { name: 'Archer', value: 'archer' },
                            { name: 'Cavalry', value: 'cavalry' },
                            { name: 'Chariot', value: 'chariot' }
                        ))
                .addIntegerOption(option => option.setName('x').setDescription('The x-coordinate.').setRequired(true))
                .addIntegerOption(option => option.setName('y').setDescription('The y-coordinate.').setRequired(true))
                .addStringOption(option =>
                    option.setName('orientation')
                        .setDescription('The direction the unit will face.')
                        .setRequired(false)
                        .addChoices(
                            { name: 'North', value: 'north' },
                            { name: 'East', value: 'east' },
                            { name: 'South', value: 'south' },
                            { name: 'West', value: 'west' }
                        )))
        .addSubcommand(subcommand =>
            subcommand
                .setName('action')
                .setDescription('Perform an action with a unit during the battle phase.')
                .addStringOption(option =>
                    option.setName('type')
                        .setDescription('The type of action to perform.')
                        .setRequired(true)
                        .addChoices(
                            { name: 'Move', value: 'move' },
                            { name: 'Turn', value: 'turn' },
                            { name: 'End Turn', value: 'end_turn' },
                            { name: 'Archer Fire', value: 'archer_fire' }
                        ))
                .addIntegerOption(option => option.setName('from_x').setDescription('The starting x-coordinate of your unit.'))
                .addIntegerOption(option => option.setName('from_y').setDescription('The starting y-coordinate of your unit.'))
                .addIntegerOption(option => option.setName('to_x').setDescription('The destination x-coordinate (for moving).'))
                .addIntegerOption(option => option.setName('to_y').setDescription('The destination y-coordinate (for moving).'))
                .addStringOption(option =>
                    option.setName('orientation')
                        .setDescription('The new orientation (for turning).')
                        .addChoices(
                            { name: 'North', value: 'north' },
                            { name: 'East', value: 'east' },
                            { name: 'South', value: 'south' },
                            { name: 'West', value: 'west' }
                        ))),

    async execute(interaction) {
        const game = gameManager.getGame(interaction.channelId);
        if (!game) {
            return interaction.reply({ content: 'This command can only be used inside a war ticket.', ephemeral: true });
        }

        const subcommand = interaction.options.getSubcommand();

        if (subcommand === 'start') {
            const x = interaction.options.getInteger('x');
            const y = interaction.options.getInteger('y');

            const result = game.startBattle(x, y);

            if (!result.success) {
                return interaction.reply({ content: result.message, ephemeral: true });
            }

            const battle = result.battle;
            const boardString = battle.renderBoard();

            const embed = new EmbedBuilder()
                .setTitle('Battle Started!')
                .setDescription(`A battle has begun at tile (${x}, ${y}).\n\n**Placement Phase**\nIt is now ${game.aggressor.username}'s turn to place their units.`)
                .addFields({ name: 'Battlefield', value: '`' + boardString + '`' });

            await interaction.reply({ embeds: [embed] });
        }
        else if (subcommand === 'place') {
            if (!game.battle) {
                return interaction.reply({ content: 'There is no battle in progress.', ephemeral: true });
            }

            const unitType = interaction.options.getString('unit_type');
            const x = interaction.options.getInteger('x');
            const y = interaction.options.getInteger('y');
            const orientation = interaction.options.getString('orientation') || 'north';

            const result = game.battle.placeUnit(interaction.user.id, unitType, x, y, orientation);

            if (!result.success) {
                return interaction.reply({ content: result.message, ephemeral: true });
            }

            const boardString = game.battle.renderBoard();
            const embed = new EmbedBuilder()
                .setTitle('Unit Placed!')
                .setDescription(result.message)
                .addFields({ name: 'Battlefield', value: '`' + boardString + '`' });

            await interaction.reply({ embeds: [embed] });
        }
        else if (subcommand === 'action') {
            if (!game.battle) {
                return interaction.reply({ content: 'There is no battle in progress.', ephemeral: true });
            }

            const actionType = interaction.options.getString('type');
            let result;

            if (actionType === 'move') {
                const fromX = interaction.options.getInteger('from_x');
                const fromY = interaction.options.getInteger('from_y');
                const toX = interaction.options.getInteger('to_x');
                const toY = interaction.options.getInteger('to_y');
                result = game.battle.moveUnit(interaction.user.id, fromX, fromY, toX, toY);
            } else if (actionType === 'turn') {
                const fromX = interaction.options.getInteger('from_x');
                const fromY = interaction.options.getInteger('from_y');
                const orientation = interaction.options.getString('orientation');
                result = game.battle.turnUnit(interaction.user.id, fromX, fromY, orientation);
            } else if (actionType === 'end_turn') {
                result = game.battle.endTurn(interaction.user.id);
            } else if (actionType === 'archer_fire') {
                const fromX = interaction.options.getInteger('from_x');
                const fromY = interaction.options.getInteger('from_y');
                const toX = interaction.options.getInteger('to_x');
                const toY = interaction.options.getInteger('to_y');
                result = game.battle.archerFire(interaction.user.id, fromX, fromY, toX, toY);
            }

            if (!result || !result.success) {
                return interaction.reply({ content: result ? result.message : "Invalid action.", ephemeral: true });
            }

            const boardString = game.battle.renderBoard();
            const embed = new EmbedBuilder()
                .setTitle('Action Taken!')
                .setDescription(result.message)
                .addFields({ name: 'Battlefield', value: '`' + boardString + '`' });

            await interaction.reply({ embeds: [embed] });
        }
    },
};

File: cogs/battle.py
import discord
from discord.ext import commands
from discord import app_commands
from game.game_manager import game_manager
from utils.battlefield_renderer import BattlefieldRenderer
from typing import Optional


class Battle(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name="battle_create_thread", description="Create a new battle thread.")
    @app_commands.describe(
        opponent="The user you want to battle against",
        thread_name="Optional custom name for the battle thread"
    )
    async def battle_create_thread(
        self,
        interaction: discord.Interaction,
        opponent: discord.Member,
        thread_name: Optional[str] = None
    ):
        if interaction.user.id == opponent.id:
            return await interaction.response.send_message(
                "You cannot start a battle with yourself!",
                ephemeral=True
            )

        if not thread_name:
            thread_name = (
                f"Battle: {interaction.user.display_name} vs "
                f"{opponent.display_name}"
            )

        # Create the thread
        thread = await interaction.channel.create_thread(  # type: ignore[reportCallIssue]
            name=thread_name,
            type=discord.ChannelType.public_thread  # type: ignore[reportCallIssue]
        )

        # Add both players to the thread
        await thread.add_user(interaction.user)
        await thread.add_user(opponent)

        # Create the game instance
        aggressor = {'id': interaction.user.id, 'name': interaction.user.display_name}
        defender = {'id': opponent.id, 'name': opponent.display_name}
        game = game_manager.create_game(thread.id, aggressor, defender)

        if not game:
            await thread.delete()
            return await interaction.response.send_message(
                "Failed to create battle thread. A game might already exist.",
                ephemeral=True
            )

        # Create initial armies for both players
        game.add_army(interaction.user.id)
        game.add_army(opponent.id)

        embed = discord.Embed(
            title="⚔️ Battle Thread Created!",
            description=(
                f"**{interaction.user.display_name}** challenges "
                f"**{opponent.display_name}** to battle!\n\n"
                f"**Initial Setup:**\n"
                f"• Both players have been given Army #1\n"
                f"• Each army starts with: 5 infantry, 1 commander\n\n"
                f"**Next Steps:**\n"
                f"• Use `/army_view` to see your armies\n"
                f"• Use `/army_modify` to customize your armies\n"
                f"• Use `/battle_start` to begin the battle\n"
                f"• INFANTRY = 5 Infantry Units (Starter)"
                f"• SHOCK = 3 Shock Units (1 Bronze)"
                f"• ARCHER = 3 Archer Units (1 Timber)"
                f"• COMMANDER = 1 Commander (Starter)"
                f"• CAVALRY = 4 Cavalry Units (1 Mount)"
                f"• CHARIOT = 2 Chariot Units (1 Mount)"
            ),
            color=discord.Color.red()
        )

        await interaction.response.send_message(
            f"Battle thread created: {thread.mention}",
            ephemeral=True
        )
        await thread.send(embed=embed)

    @app_commands.command(name="battle_start", description="Start a battle between your armies.")
    @app_commands.describe(
        aggressor_army="ID of the aggressor's army",
        defender_army="ID of the defender's army"
    )
    async def battle_start(
        self,
        interaction: discord.Interaction,
        aggressor_army: int,
        defender_army: int
    ):
        game = game_manager.get_game(interaction.channel_id)
        if not game:
            return await interaction.response.send_message(
                "This command can only be used inside a battle thread.",
                ephemeral=True
            )

        if interaction.user.id != game.aggressor['id']:
            return await interaction.response.send_message(
                "Only the aggressor can start battles.",
                ephemeral=True
            )

        result = game.start_battle(aggressor_army, defender_army)
        if not result['success']:
            return await interaction.response.send_message(result['message'], ephemeral=True)

        renderer = BattlefieldRenderer(game.battle.board)
        image = renderer.render_board()

        embed = discord.Embed(
            title="Battle Started!",
            description=(
                f"A battle has begun!\n\n**Placement Phase**\n"
                f"It is now {game.aggressor['name']}'s turn to place "
                f"their units."
            )
        )
        embed.set_image(url="attachment://battlefield.png")

        await interaction.response.send_message(
            embed=embed,
            file=discord.File(image, filename="battlefield.png")
        )

    @app_commands.command(name="battle_place", description="Place a unit on the battlefield.")
    @app_commands.describe(
        unit_type="Type of unit to place",
        x="X coordinate (0-8)",
        y="Y coordinate (0-8)",
        orientation="Direction the unit should face"
    )
    @app_commands.choices(unit_type=[
        app_commands.Choice(name="Infantry", value="infantry"),
        app_commands.Choice(name="Commander", value="commander"),
        app_commands.Choice(name="Shock", value="shock"),
        app_commands.Choice(name="Archer", value="archer"),
        app_commands.Choice(name="Cavalry", value="cavalry"),
        app_commands.Choice(name="Chariot", value="chariot")
    ])
    @app_commands.choices(orientation=[
        app_commands.Choice(name="North", value="north"),
        app_commands.Choice(name="South", value="south"),
        app_commands.Choice(name="East", value="east"),
        app_commands.Choice(name="West", value="west")
    ])
    async def battle_place(
        self,
        interaction: discord.Interaction,
        unit_type: str,
        x: int,
        y: int,
        orientation: str = "north"
    ):
        game = game_manager.get_game(interaction.channel_id)
        if not game or not game.battle:
            return await interaction.response.send_message(
                "There is no battle in progress.",
                ephemeral=True
            )

        result = game.battle.place_unit(
            interaction.user.id, unit_type, x, y, orientation
        )
        if not result['success']:
            return await interaction.response.send_message(result['message'], ephemeral=True)

        renderer = BattlefieldRenderer(game.battle.board)
        image = renderer.render_board()

        embed = discord.Embed(
            title="Unit Placed!",
            description=result['message']
        )
        embed.set_image(url="attachment://battlefield.png")

        await interaction.response.send_message(
            embed=embed,
            file=discord.File(image, filename="battlefield.png")
        )

    @app_commands.command(name="battle_action", description="Perform a battle action.")
    @app_commands.describe(
        action_type="Type of action to perform",
        from_x="Starting X coordinate",
        from_y="Starting Y coordinate",
        to_x="Target X coordinate",
        to_y="Target Y coordinate",
        orientation="Direction to face"
    )
    @app_commands.choices(action_type=[
        app_commands.Choice(name="Move", value="move"),
        app_commands.Choice(name="Turn", value="turn"),
        app_commands.Choice(name="End Turn", value="end_turn")
    ])
    async def battle_action(
        self,
        interaction: discord.Interaction,
        action_type: str,
        from_x: Optional[int] = None,
        from_y: Optional[int] = None,
        to_x: Optional[int] = None,
        to_y: Optional[int] = None,
        orientation: Optional[str] = None
    ):
        game = game_manager.get_game(interaction.channel_id)
        if not game or not game.battle:
            return await interaction.response.send_message(
                "There is no battle in progress.",
                ephemeral=True
            )

        result = None
        if action_type == 'move':
            result = game.battle.move_unit(
                interaction.user.id, from_x, from_y, to_x, to_y
            )
        elif action_type == 'turn':
            result = game.battle.turn_unit(
                interaction.user.id, from_x, from_y, orientation
            )
        elif action_type == 'end_turn':
            result = game.battle.end_turn(interaction.user.id)
        # ... other actions ...

        if not result or not result['success']:
            return await interaction.response.send_message(
                result['message'] if result else "Invalid action.",
                ephemeral=True
            )

        if result.get('battle_ended'):
            winner_user = (
                game.aggressor if result['winner'] == game.aggressor['id']
                else game.defender
            )
            embed = discord.Embed(
                title='🏆 Battle Concluded!',
                description=(
                    f"**{winner_user['name']} has won the battle!**\n\n"
                    f"{result['message']}"
                ),
                color=(
                    discord.Color.red()
                    if result['winner'] == game.aggressor['id']
                    else discord.Color.blue()
                )
            )
            game_manager.end_battle(interaction.channel_id)
        else:
            embed = discord.Embed(
                title="Action Taken!",
                description=result['message']
            )

        renderer = BattlefieldRenderer(game.battle.board)
        image = renderer.render_board()
        embed.set_image(url="attachment://battlefield.png")

        await interaction.response.send_message(
            embed=embed,
            file=discord.File(image, filename="battlefield.png")
        )

    @app_commands.command(name="battle_forfeit", description="Forfeit the current battle.")
    async def battle_forfeit(self, interaction: discord.Interaction):
        game = game_manager.get_game(interaction.channel_id)
        if not game or not game.battle:
            return await interaction.response.send_message(
                "There is no battle in progress.",
                ephemeral=True
            )

        result = game.battle.forfeit_battle(interaction.user.id)
        if not result['success']:
            return await interaction.response.send_message(result['message'], ephemeral=True)

        game_manager.end_battle(interaction.channel_id)
        # lock the thread to prevent further messages
        try:
            await interaction.channel.edit(locked=True)  # type: ignore
        except Exception:
            pass
        await interaction.response.send_message(result['message'])


async def setup(bot):
    await bot.add_cog(Battle(bot))
